Fast startup

spencerkimball opened the pull request.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/28581)
<!-- Reviewable:end -->

spencerkimball committed:
storage: replace Replica objects with lazily-loaded ReplicaShim

Initializing all replicas in memory at startup is costly. For example,
65,000 replicas takes 20+ seconds on my laptop. Further, the computation
of metrics every 10s requires all 65,000 replicas to contribute, and
shows up prominently on traces and no doubt is contributing to long
tail latencies.

This change replaces the use of naked `*Replica` objects in `Store`
with a new `ReplicaShim` struct containing either an initialized
`*Replica*`, or else the `RangeDescriptor` and other data required
to compute metrics cheaply.

Release note (performance improvement): for nodes with many replicas,
startup time and metrics computation should be considerably better /
more efficient.

spencerkimball committed:
storage: replace Replica objects with lazily-loaded ReplicaShim

Initializing all replicas in memory at startup is costly. For example,
65,000 replicas takes 20+ seconds on my laptop. Further, the computation
of metrics every 10s requires all 65,000 replicas to contribute, and
shows up prominently on traces and no doubt is contributing to long
tail latencies.

This change replaces the use of naked `*Replica` objects in `Store`
with a new `ReplicaShim` struct containing either an initialized
`*Replica*`, or else the `RangeDescriptor` and other data required
to compute metrics cheaply.

Release note (performance improvement): for nodes with many replicas,
startup time and metrics computation should be considerably better /
more efficient.

bobvawter commented:
[![CLA assistant check](https://cla.crdb.dev/pull/badge/not_signed)](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=28581) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=28581) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.crdb.dev/check/cockroachdb/cockroach?pullRequest=28581) it.</sub>
